File: src/test/java/im/tox/tox4j/v2/callbacks/FileTransferTest.java
package im.tox.tox4j.v2.callbacks;

import im.tox.tox4j.v2.AliceBobTestBase;
import im.tox.tox4j.v2.ToxCore;
import im.tox.tox4j.v2.ToxCoreImpl;
import im.tox.tox4j.v2.ToxOptions;
import im.tox.tox4j.v2.enums.ToxFileControl;
import im.tox.tox4j.v2.enums.ToxFileKind;
import im.tox.tox4j.v2.exceptions.SpecificToxException;
import im.tox.tox4j.v2.exceptions.ToxNewException;

import java.util.Arrays;
import java.util.Random;

import static org.junit.Assert.*;

public class FileTransferTest extends AliceBobTestBase {

    @Override
    protected ToxCore newTox(ToxOptions options) throws ToxNewException {
        return new ToxCoreImpl(options);
    }

    @Override
    protected ChatClient newClient() {
        return new Client();
    }


    private static class Client extends ChatClient {

        private static final byte[] fileData = new byte[1500];
        static {
            new Random().nextBytes(fileData);
        }

        private final byte[] receivedData = new byte[fileData.length];
        private long position = 0;
        private Integer sentFileNumber = null;

        @Override
        public void friendConnected(final int friendNumber, boolean isConnected) {
            debug("is now connected to friend " + friendNumber);
            assertEquals(0, friendNumber);
            if (isBob()) return;
            addTask(new Task() {
                @Override
                public void perform(ToxCore tox) throws SpecificToxException {
                    sentFileNumber = tox.fileSend(friendNumber, ToxFileKind.DATA, fileData.length,
                            ("file for " + getFriendName() + ".png").getBytes());
                }
            });
        }

        @Override
        public void fileReceive(final int friendNumber, final int fileNumber, ToxFileKind kind, long fileSize, byte[] filename) {
            debug("received file send request " + fileNumber + " from friend number " + friendNumber);
            assertTrue(isBob());
            assertEquals(0, friendNumber);
            assertEquals(0 | 0x100, fileNumber);
            assertEquals(ToxFileKind.DATA, kind);
            assertEquals(fileData.length, fileSize);
            assertEquals("file for " + getName() + ".png", new String(filename));
            addTask(new Task() {
                @Override
                public void perform(ToxCore tox) throws SpecificToxException {
                    debug("sending control RESUME for " + fileNumber);
                    tox.fileControl(friendNumber, fileNumber, ToxFileControl.RESUME);
                }
            });
        }

        @Override
        public void fileControl(int friendNumber, int fileNumber, ToxFileControl control) {
            debug("file control from " + friendNumber + " for file " + fileNumber + ": " + control);
            assertTrue(isAlice());
        }

        @Override
        public void fileSendChunk(final int friendNumber, final int fileNumber, final long position, final int length) {
            debug("got request for " + length + "B from " + friendNumber + " for file " + fileNumber + " at " + position);
            assertTrue(isAlice());
            assertTrue(position >= 0);
            assertTrue(position < Integer.MAX_VALUE);
            assertEquals(sentFileNumber.intValue(), fileNumber);
            if (length == 0) {
                // Sending is done, clear sending status.
                sentFileNumber = null;
                finish();
                return;
            }
            addTask(new Task() {
                @Override
                public void perform(ToxCore tox) throws SpecificToxException {
                    debug("sending " + length + "B to " + friendNumber);
                    tox.fileSendChunk(friendNumber, fileNumber,
                            Arrays.copyOfRange(fileData, (int) position, Math.min((int) position + length, fileData.length)));
                }
            });
        }

        @Override
        public void fileReceiveChunk(int friendNumber, int fileNumber, long position, byte[] data) {
            debug("got " + data.length + "B from " + friendNumber + " at " + position);
            assertTrue(isBob());
            assertEquals(0, friendNumber);
            assertEquals(0 | 0x100, fileNumber);
            assertEquals(this.position, position);
            assertNotNull(data);
            assertNotEquals(0, data.length);
            this.position += data.length;
            assertTrue(this.position <= receivedData.length);
            System.arraycopy(data, 0, receivedData, (int) position, data.length);
            if (this.position == receivedData.length) {
                assertArrayEquals(fileData, receivedData);
                finish();
            }
        }
    }

}

File: projects/tox4j/src/test/java/im/tox/tox4j/core/callbacks/FileTransferTest.scala
package im.tox.tox4j.core.callbacks

import java.util.Random

import im.tox.tox4j.core.enums.{ToxConnection, ToxFileControl, ToxFileKind}
import im.tox.tox4j.testing.autotest.{AliceBobTest, AliceBobTestBase}

final class FileTransferTest extends AliceBobTest {

  private val fileData = new Array[Byte](1500)
  new Random().nextBytes(fileData)

  sealed case class State(
      receivedData: Array[Byte] = Array.ofDim[Byte](fileData.length),
      position: Long = 0L,
      sentFileNumber: Int = -1
  ) {
    def addData(position: Long, data: Array[Byte]): State = {
      assert(data.nonEmpty)
      val nextPosition = this.position + data.length
      assert(nextPosition <= this.receivedData.length)
      System.arraycopy(data, 0, this.receivedData, position.toInt, data.length)
      copy(position = nextPosition)
    }
  }

  override def initialState: State = State()

  protected override def newChatClient(name: String, expectedFriendName: String) = new Alice(name, expectedFriendName)

  class Alice(name: String, expectedFriendName: String) extends ChatClient(name, expectedFriendName) {

    override def friendConnectionStatus(friendNumber: Int, connectionStatus: ToxConnection)(state: ChatState): ChatState = {
      super.friendConnectionStatus(friendNumber, connectionStatus)(state)
      if (connectionStatus != ToxConnection.NONE && isAlice) {
        state.addTask { (tox, state) =>
          val sentFileNumber = tox.fileSend(
            friendNumber,
            ToxFileKind.DATA,
            fileData.length,
            Array.ofDim[Byte](0),
            s"file for $expectedFriendName.png".getBytes
          )
          state.set(state.get.copy(sentFileNumber = sentFileNumber))
        }
      } else {
        state
      }
    }

    override def fileRecv(friendNumber: Int, fileNumber: Int, kind: Int, fileSize: Long, filename: Array[Byte])(state: ChatState): ChatState = {
      debug(s"received file send request $fileNumber from friend number $friendNumber")
      assert(isBob)
      assert(friendNumber == AliceBobTestBase.FriendNumber)
      assert(fileNumber == (0 | 0x10000))
      assert(kind == ToxFileKind.DATA)
      assert(fileSize == fileData.length)
      assert(new String(filename) == s"file for $name.png")
      state.addTask { (tox, state) =>
        debug("sending control RESUME for " + fileNumber)
        tox.fileControl(friendNumber, fileNumber, ToxFileControl.RESUME)
        state
      }
    }

    override def fileRecvControl(friendNumber: Int, fileNumber: Int, control: ToxFileControl)(state: ChatState): ChatState = {
      debug(s"file control from $friendNumber for file $fileNumber: $control")
      assert(isAlice)
      state
    }

    override def fileChunkRequest(friendNumber: Int, fileNumber: Int, position: Long, length: Int)(state: ChatState): ChatState = {
      debug(s"got request for ${length}B from $friendNumber for file $fileNumber at $position")
      assert(friendNumber == AliceBobTestBase.FriendNumber)
      assert(isAlice)
      assert(position >= 0)
      assert(position < Int.MaxValue)
      assert(fileNumber == state.get.sentFileNumber)
      if (length == 0) {
        state.set(state.get.copy(sentFileNumber = -1)).finish
      } else {
        state.addTask { (tox, state) =>
          debug(s"sending ${length}B to $friendNumber")
          tox.fileSendChunk(
            friendNumber,
            fileNumber,
            position,
            fileData.slice(
              position.toInt,
              Math.min(position.toInt + length, fileData.length)
            )
          )
          state
        }
      }
    }

    override def fileRecvChunk(friendNumber: Int, fileNumber: Int, position: Long, data: Array[Byte])(state: ChatState): ChatState = {
      debug(s"got ${data.length}B from $friendNumber at $position")
      assert(isBob)
      assert(friendNumber == AliceBobTestBase.FriendNumber)
      assert(fileNumber == (0 | 0x10000))
      assert(position == state.get.position)
      assert(data != null)
      if (state.get.position == state.get.receivedData.length) {
        assert(data.isEmpty)
        assert(state.get.receivedData sameElements fileData)
        state.finish
      } else {
        state.set(state.get.addData(position, data))
      }
    }
  }

}
